include/openssl/asn1t.h

```diff
@@ -880,7 +880,11 @@ DECLARE_ASN1_ITEM(LONG)
 DECLARE_ASN1_ITEM(ZLONG)
 # endif
 
+#if defined(OPENSSL_BUILDING_OPENSSL) || !defined(OPENSSL_NO_DEPRECATED_3_0)
+STACK_OF(ASN1_VALUE);
+#else
 DEFINE_STACK_OF(ASN1_VALUE)
+#endif
```

Comment:
So yeahok, interesting construct...  but not at all sure what you want to achieve.  The only condition when the DEFINE macros will be used is when an external application uses this header, and `no-deprecated` has been configured (which it shouldn't for generic system installs).

Since you have added the DEFINE macro calls in the appropriate .c files, the `OPENSSL_BUILDING_OPENSSL` macro check (in fact, the macro itself) has become irrelevant (*thankfully*), so it can be removed.  This leaves `no-deprecated`...  but it's *weird* that the full definition is defined here *in a configuration that removes public defs*!

All in all, I was rather expecting to see all `DEFINE_STACK_OF` calls replaced with `STACK_OF` calls.  For all intents and purposes, that's what you've done, but just with some extra fluff that blurs the eye.

Interestingly enough, you have done said replacement in some header files, such as `bio.h`.  That's exactly what I expected to see.

Reply:
I don't think I understand your comments, or you don't understand what I did. :)

The existing public header files, when used publicly, have to have the full definitions of all inline functions. If no-deprecated is configured, or when the header file is used for internal builds of OpenSSL, we don't want the inline functions in the header file.  As a result, the header file works for three cases:

- Building openssl itself, which doesnt want the functions defined in the header file, no matter how it is configured
- Having the existing inline functions defined in the header file (configured without no-deprecated)
- Not having the inline functions defined, just the STACK type (when configured with no-deprecated)

The change in bio.h is incomplete, it should use the if/else/endif construct like is used asn1.h to either just declare the type or define the functions.  Just declaring the datatype, like bio.h, doesn't meet all three criteria mentioned above and breaks the existing, albeit implied, API.

I'll do another PR to fix all the rest of the headers soon.

Reply:
OK, so if that's what you want to achieve, I believe that your logic is a little bit flawed. Easy to change, though 

``` C
-#if defined(OPENSSL_BUILDING_OPENSSL) || !defined(OPENSSL_NO_DEPRECATED_3_0)
+#if defined(OPENSSL_BUILDING_OPENSSL) || defined(OPENSSL_NO_DEPRECATED_3_0)
```

Reply:
I can't stand those double-negatives. :) I know why, but still it's very confusing.  Thanks.

Reply:
Taking yet another look, I think the current code is right and the proposed change is wrong.  We want to say "if building openssl or no-deprecated, just declare" which is what the code currently does.

I have another idea for simplication which is to put this in safestack.h
```
#if defined(OPENSSL_BUILD_OPENSSL) || !defined(OPENSSL_NO_DEPRECATED_3_0)
# define DEFINE_OR_DECLARE_STACK_OF(s) STACK_OF(s)
#else 
# define DEFINE_OR_DECLARE_STACK_OF(s) DEFINE_STACK_OF(s)
#endif 
```
And the edit the header files to juse use ```DEFINE_OR_DECLARE```

Reply:
The condition you currently have says "if building openssl or *not* no-deprecated, then just declare"

Reply:
I'll double check but as I read thigns the no-deprecated flag says "do not deprecate things"  So it's a triple-negative.

Reply:
okay, looking around you're right and i'm wrong.

Reply:
Reading the docs often helps :wink:

https://github.com/openssl/openssl/blob/b0d3442efc10b635863b915c2d014345f6e5a219/INSTALL#L356-L360